File: pe58.c
#include <stdio.h>
#include <stdlib.h>
#define MAX 1000000000

int main() {
	char *sieve;
	int i, j, m, n, prime, diag;
	sieve = (char*)calloc(MAX/2, 1);
	puts("calloc'd");
	m = 0;
	for (i=0; i<MAX/2; i++) {
		if (i==m*m && m < 1000) {
			printf("%d..", i);
			fflush(stdout);
			m++;
		}
		if (sieve[i] == 0)
			for (j=i*3+3; j<MAX/2; j+=i*2+3)
				sieve[j] = 1;
	}
	puts("\nprime init");
	m = 2;
	n = 1;
	prime = 0;
	diag = 1;
	while (n<MAX) {
		for (i=0; i<4; i++) {
			n += m;
			if (n%2 && sieve[(n-3)/2] == 0)
				prime++;
		}
		diag += 4;
		if (10*prime < diag) {
			printf("sidelength: %d\n", m+1);
			goto end;
		}
		m += 2;
	}
	puts("n MAX'd");
end:
	free(sieve);
	return 0;
}

File: pe58.py
max = 100000000
sieve = [0 for i in xrange(max)]
for i in xrange(2, max):
	if sieve[i] == 0:
		for j in xrange(2*i, max, i):
			sieve[j] = 1

def isprime(n):
	return n>1 and sieve[n] == 0

print "prime init"

m = 2
n = 1
primecnt = 0.
diagcnt = 1
while True:
	for i in xrange(4):
		n += m
		if isprime(n): primecnt += 1
	diagcnt += 4
	if 10*primecnt < diagcnt:
		print m+1
		break
	m += 2
